File: raw/github/paypal/paypal-ios-sdk-demo-app/snapshots/2026-08-16-047a50e/files/PayPalDemo/CardCheckoutViews/CartView.swift
import SwiftUI
import PaymentButtons

struct CartView: View {

    @EnvironmentObject private var coordinator: CheckoutCoordinator

    let items: [Item]
    let isPaymentLink: Bool

    private var totalAmount: Double {
        items.reduce(0, { $0 + $1.amount})
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 30) {
            Text("Cart")
                .font(.largeTitle)
                .padding([.top, .leading])
            
            ForEach(items) { item in
                VStack {
                    CartItemView(item: item)
                }
            }

            Divider()
                .padding(.vertical)
                .padding(.horizontal)
            
            TotalSection(amount: totalAmount)

            Spacer()
            
            VStack(spacing: 10) {
                if isPaymentLink {
                    PaymentButton(
                        title: "Pay Now",
                        imageName: nil,
                        backgroundColor: Color.black,
                        action: {
                            if let url = URL(string: "https://www.sandbox.paypal.com/ncp/payment/BFXRZ54VKCAQ6") {
                                coordinator.openPaymentLink(url: url)
                            }
                        }
                    )
                } else {
                    PayPalButton.Representable(
                        size: .expanded,
                        label: .payWith
                    ){
                        coordinator.startPayPalCheckout(amount: totalAmount)
                    }
                    .fixedSize(horizontal: false, vertical: true)
                    PaymentButton(
                        title: "Pay with Card",
                        imageName: nil,
                        backgroundColor: Color.black,
                        action: {
                            coordinator.startCardCheckout(amount: totalAmount)
                        }
                    )
                }


            }
            .padding(.horizontal)
            .padding(.bottom, 40)
        }
        .background(Color(.systemBackground))
        .edgesIgnoringSafeArea(.bottom)
    }
}


struct CartItemView: View {    
    let item: Item

    var body: some View {
        VStack(alignment: .leading, spacing: 15){
            HStack(alignment: .center) {
                if let image1 = item.image {
                    image1
                        .resizable()
                        .scaledToFit()
                        .frame(width: 100, height: 100)
                        .background(Color.gray.opacity(0.2))
                        .cornerRadius(8)
                }
                VStack(alignment: .leading) {
                    Text(item.name)
                        .font(.headline)
                }
                
                Spacer()
                
                Text("$\(item.amount, specifier: "%.2f")")
                    .font(.headline)
            }
            .padding()
            .background(Color.white.opacity(0.1))
            .cornerRadius(8)
            .overlay(
                RoundedRectangle(cornerRadius: 8)
                    .stroke(Color.black, lineWidth: 1)
            )
        }
        .padding(.horizontal)
    }
    
}

struct TotalSection: View {
    let amount: Double

    var body: some View {
        HStack {
            Text("Total")
                .font(.title2)
            Spacer()
            
            Text("$\(amount, specifier: "%.2f")")
                .font(.title2)
        }
        .padding(.horizontal)
        
    }
    
}

struct PaymentButton: View {
    let title: String
    let imageName: String?
    let backgroundColor: Color
    let action: () -> Void
    
    var body: some View {
        Button(action: action) {
            HStack {
                if let imageName = imageName {
                    Image(imageName)
                        .resizable()
                        .scaledToFit()
                        .frame(width: 24, height: 24)
                }
                Text(title)
                    .bold()
            }
            .frame(maxWidth: .infinity)
            .padding()
            .foregroundColor(.white)
            .background(backgroundColor)
            .cornerRadius(4)
        }
    }
}

File: raw/github/paypal/paypal-android-sdk-demo-app/snapshots/2026-08-16-d1137d5/files/app/src/main/java/com/firstapp/paypaldemo/main/CartView.kt
package com.firstapp.paypaldemo.main

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.SegmentedButton
import androidx.compose.material3.SegmentedButtonDefaults
import androidx.compose.material3.SingleChoiceSegmentedButtonRow
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.derivedStateOf
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.compose.ui.viewinterop.AndroidView
import com.firstapp.paypaldemo.Constants.SHOPPING_CART_ITEMS
import com.paypal.android.paymentbuttons.PayPalButton
import com.paypal.android.paymentbuttons.PayPalButtonColor
import com.paypal.android.paymentbuttons.PayPalButtonLabel
import com.paypal.android.paymentbuttons.PaymentButtonSize

data class Item(
    val name: String,
    val amount: Double,
    val imageResId: Int
)

@ExperimentalMaterial3Api
@Composable
fun CartView(
    shoppingCartItems: List<Item>,
    onPayWithCard: (Double) -> Unit,
    onPayWithLink: (Double) -> Unit,
    onPayWithPayPal: () -> Unit,
) {
    // TODO: make ShoppingCart data type with .items and .totalAmount property
    val totalAmount by remember { derivedStateOf { shoppingCartItems.sumOf { it.amount } } }
    val payPalButtonCornerRadius = with(LocalDensity.current) { 10.dp.toPx() }

    var isPaymentLinkEnabled by rememberSaveable { mutableStateOf(true) }

    Column(
        modifier = Modifier
            .fillMaxSize()
            .background(MaterialTheme.colorScheme.background)
            .padding(top = 16.dp, start = 20.dp, end = 20.dp),

        verticalArrangement = Arrangement.Top,
        horizontalAlignment = Alignment.Start
    ) {
        SingleChoiceSegmentedButtonRow(modifier = Modifier.fillMaxWidth()) {
            SegmentedButton(
                selected = isPaymentLinkEnabled,
                onClick = { isPaymentLinkEnabled = true },
                shape = SegmentedButtonDefaults.itemShape(index = 0, 2)
            ) {
                Text("Use Payment Link")
            }
            SegmentedButton(
                selected = !isPaymentLinkEnabled,
                onClick = { isPaymentLinkEnabled = false },
                shape = SegmentedButtonDefaults.itemShape(index = 1, 2)
            ) {
                Text("Use PayPal SDK")
            }
        }
        Text(
            text = "Cart",
            fontSize = 24.sp,
            fontWeight = FontWeight.Bold,
            modifier = Modifier.padding(top = 8.dp, bottom = 25.dp)
        )

        shoppingCartItems.forEach { item ->
            CartItemView(item)
        }

        HorizontalDivider(
            modifier = Modifier
                .padding(vertical = 16.dp)
                .fillMaxWidth(),
            color = Color.Gray.copy(alpha = 0.2f)
        )

        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.SpaceBetween
        ) {
            Text(
                text = "Total",
                fontSize = 20.sp,
                fontWeight = FontWeight.Medium
            )
            Text(
                text = "$${"%.2f".format(totalAmount)}",
                fontSize = 20.sp,
                fontWeight = FontWeight.Medium
            )
        }

        Spacer(modifier = Modifier.weight(1f))
        if (isPaymentLinkEnabled) {
            PaymentButton(
                text = "Pay Now",
                backgroundColor = Color.Black,
                onClick = {
                    onPayWithLink(totalAmount)
                }
            )
        } else {
            PayPalButton(
                cornerRadius = payPalButtonCornerRadius,
                modifier = Modifier.fillMaxWidth(),
                onClick = { onPayWithPayPal() }
            )
            Spacer(modifier = Modifier.height(10.dp))
            PaymentButton(
                text = "Pay with Card",
                backgroundColor = Color.Black,
                onClick = { onPayWithCard(totalAmount) }
            )
        }
    }
}

@Composable
fun CartItemView(item: Item) {
    Box(
        modifier = Modifier
            .fillMaxWidth()
    ) {
        Surface(
            modifier = Modifier
                .fillMaxWidth(),
            shape = RoundedCornerShape(8.dp),
            border = BorderStroke(1.dp, Color.Black),
        ) {
            Row(
                modifier = Modifier
                    .padding(16.dp),
                verticalAlignment = Alignment.CenterVertically
            ) {
                Image(
                    painter = painterResource(id = item.imageResId),
                    contentDescription = item.name,
                    modifier = Modifier
                        .size(50.dp)
                        .clip(RoundedCornerShape(8.dp))
                        .background(Color.Gray.copy(alpha = 0.2f))
                )

                Spacer(modifier = Modifier.width(16.dp))

                Text(
                    text = item.name,
                    fontWeight = FontWeight.Bold,
                    modifier = Modifier.weight(1f)
                )

                Text(
                    text = "$${"%.2f".format(item.amount)}",
                    fontWeight = FontWeight.Bold
                )
            }
        }
    }
}

@Composable
fun PayPalButton(
    modifier: Modifier = Modifier,
    cornerRadius: Float? = null,
    onClick: () -> Unit
) {
    AndroidView(
        factory = { context ->
            PayPalButton(context).apply { setOnClickListener { onClick() } }
        },
        update = { button ->
            button.color = PayPalButtonColor.BLUE
            button.label = PayPalButtonLabel.PAY
            button.size = PaymentButtonSize.LARGE
            button.customCornerRadius = cornerRadius
        },
        modifier = modifier
    )
}

@Composable
fun PaymentButton(
    text: String,
    backgroundColor: Color,
    onClick: () -> Unit
) {
    Button(
        onClick = onClick,
        modifier = Modifier
            .fillMaxWidth()
            .height(50.dp),
        colors = ButtonDefaults.buttonColors(containerColor = backgroundColor),
        shape = RoundedCornerShape(10.dp)
    ) {
        Row(
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.Center
        ) {
            Text(
                text = text,
                fontSize = 18.sp,
                fontWeight = FontWeight.Bold,
                color = Color.White
            )
        }
    }
}

@ExperimentalMaterial3Api
@Preview
@Composable
fun CartViewPreview() {
    MaterialTheme {
        Surface(modifier = Modifier.fillMaxSize()) {
            CartView(
                onPayWithCard = {},
                onPayWithPayPal = {},
                shoppingCartItems = SHOPPING_CART_ITEMS,
                onPayWithLink = {}
            )
        }
    }
}
